### PR TITLE
Add callback in FileSingleStreamSpiller to clear Factory cache on disk error

### PR DESCRIPTION
**Pull request by dfinninger**
Follow up work to #2444 

Not married to the callback structure. It seemed the most efficient way to get it done without refactoring a bunch of code. If a better way comes to mind, let me know.

### CODE REVIEW

**Comment by kokosing** [comment]
@cla-bot check

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java [hunk 1]
```diff
@@ -215,4 +225,8 @@ private boolean isSeeminglyHealthy(Path path)
             return false;
         }
     }
+
+    private void clearDiskHealthCache() {
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: code style mistake. Other than that I would inline this method.

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java [hunk 2]
```diff
@@ -78,7 +78,9 @@
     private final double maxUsedSpaceThreshold;
     private final boolean spillEncryptionEnabled;
     private int roundRobinIndex;
-    private final LoadingCache<Path, Boolean> spillPathHealthCache;
+
+    @VisibleForTesting
+    final LoadingCache<Path, Boolean> spillPathHealthCache;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: Maybe instead of exposing a field, create a method that exposes just the size of cache.

**Comment by dfinninger** [comment]
@kokosing I've addressed comments and checks.

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java [hunk 3]
```diff
@@ -99,10 +102,12 @@ public FileSingleStreamSpiller(
         // before/after the spiller thread allocates that memory -- -- whether before or after depends on whether writePages() is in the
         // middle of execution when close() is called (note that this applies to both readPages() and writePages() methods).
         this.memoryContext.setBytes(BUFFER_SIZE);
+        this.filesystemErrorHandler = filesystemErrorHandler;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java]: requireNonNull

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java (hunk 2 above)
Comment (kokosing) [reply presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: It looks like this comment is not addressed

Comment (dfinninger) [reply presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: Ah, didn't push the last commit. Fixed.

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java [hunk 4]
```diff
@@ -99,10 +102,12 @@ public FileSingleStreamSpiller(
         // before/after the spiller thread allocates that memory -- -- whether before or after depends on whether writePages() is in the
         // middle of execution when close() is called (note that this applies to both readPages() and writePages() methods).
         this.memoryContext.setBytes(BUFFER_SIZE);
+        this.filesystemErrorHandler = requireNonNull(filesystemErrorHandler);
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java]: add a message that `"filesystemErrorHandler is null"`

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java [hunk 5]
```diff
@@ -71,14 +71,17 @@
     private long spilledPagesInMemorySize;
     private ListenableFuture<?> spillInProgress = Futures.immediateFuture(null);
 
+    private Runnable filesystemErrorHandler;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java]: final

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java [hunk 6]
```diff
@@ -78,6 +78,7 @@
     private final double maxUsedSpaceThreshold;
     private final boolean spillEncryptionEnabled;
     private int roundRobinIndex;
+
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: undo

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java [hunk 7]
```diff
@@ -215,4 +224,9 @@ private boolean isSeeminglyHealthy(Path path)
             return false;
         }
     }
+
+    protected long getSpillPathCacheSize()
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: `@VisibleForTesting`?

File: presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java [hunk 8]
```diff
@@ -147,6 +147,87 @@ public void testDistributesSpillOverPathsBadDisk()
         assertEquals(listFiles(spillPath2.toPath()).size(), 0);
     }
 
+    @Test
+    public void testCacheInvalidatedOnBadDisk()
+            throws Exception
+    {
+        List<Type> types = ImmutableList.of(BIGINT);
+        List<Path> spillPaths = ImmutableList.of(spillPath1.toPath(), spillPath2.toPath());
+        FileSingleStreamSpillerFactory spillerFactory = new FileSingleStreamSpillerFactory(
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java]: can you extract a method where you create `spillerFactory`

File: presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java [hunk 9]
```diff
@@ -147,6 +147,87 @@ public void testDistributesSpillOverPathsBadDisk()
...
+            if (i == 1) {
+                // Set second spiller path to read-only after initialization to emulate a disk failing during runtime
+                setPosixFilePermissions(spillPath2.toPath(), ImmutableSet.of(PosixFilePermission.OWNER_READ));
+            }
+
+            try {
+                getUnchecked(singleStreamSpiller.spill(page));
+            }
+            catch (Exception ignored) {
+                // Do nothing because we expect this to fail due to file permissions
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java]: Please use `org.assertj.core.api.AssertionsForClassTypes#assertThatThrownBy(org.assertj.core.api.ThrowableAssert.ThrowingCallable)` 
 to verify expected exceptions.

File: presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java [hunk 10]
```diff
@@ -147,6 +147,87 @@ public void testDistributesSpillOverPathsBadDisk()
...
+                1.0,
+                false,
+                false);
+
+        assertEquals(listFiles(spillPath1.toPath()).size(), 0);
+        assertEquals(listFiles(spillPath2.toPath()).size(), 0);
+
+        Page page = buildPage();
+        List<SingleStreamSpiller> spillers = new ArrayList<>();
+        for (int i = 0; i < 2; ++i) {
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java]: Please unroll this loop

File: presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java [hunk 11]
```diff
@@ -147,6 +147,87 @@ public void testDistributesSpillOverPathsBadDisk()
...
+            SingleStreamSpiller singleStreamSpiller = spillerFactory.create(types, bytes -> {}, newSimpleAggregatedMemoryContext().newLocalMemoryContext("test"));
+            getUnchecked(singleStreamSpiller.spill(page));
+            spillers.add(singleStreamSpiller);
+        }
+
+        assertEquals(spillerFactory.getSpillPathCacheSize(), 2, "cache contains no entries");
+
+        // restore permissions to allow cleanup
+        setPosixFilePermissions(spillPath2.toPath(), ImmutableSet.of(PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_WRITE, PosixFilePermission.OWNER_EXECUTE));
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java]: you haven't modified permissions in this test.

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java [hunk 12]
```diff
@@ -215,4 +223,10 @@ private boolean isSeeminglyHealthy(Path path)
             return false;
         }
     }
+
+    @VisibleForTesting
+    protected long getSpillPathCacheSize()
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpillerFactory.java]: I think it does need to be `protected`. Just `package` should be enough.

File: presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java [hunk 13]
```diff
@@ -75,20 +76,30 @@ public void tearDown()
...
-            throws Exception
+    private FileSingleStreamSpillerFactory spillerFactoryFactory(List<Path> paths)
     {
-        List<Type> types = ImmutableList.of(BIGINT);
-        List<Path> spillPaths = ImmutableList.of(spillPath1.toPath(), spillPath2.toPath());
-        FileSingleStreamSpillerFactory spillerFactory = new FileSingleStreamSpillerFactory(
+        return spillerFactoryFactory (paths, 1.0);
+    }
+
+    private FileSingleStreamSpillerFactory spillerFactoryFactory(List<Path> paths, Double maxUsedSpaceThreshold)
```
Comment (kokosing) [inline presto-main/src/test/java/io/prestosql/spiller/TestFileSingleStreamSpillerFactory.java]: please extract this change to separate commit, before your actual change. Because it is just a refactoring. Also please put this method somewhere at the bottom, it is not just much important as it is only utility method.

**Comment by dfinninger** [comment]
@kokosing this should be ready for review again.
It was a busy week at work so it took me a while to get around to.

**Comment by dfinninger** [comment]
@kokosing Was there anything else you wanted to see in this commit or is this good to merge?

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java [hunk 14]
```diff
@@ -71,14 +71,17 @@
...
     public FileSingleStreamSpiller(
             PagesSerde serde,
             ListeningExecutorService executor,
             Path spillPath,
             SpillerStats spillerStats,
             SpillContext spillContext,
             LocalMemoryContext memoryContext,
-            Optional<SpillCipher> spillCipher)
+            Optional<SpillCipher> spillCipher,
+            Runnable filesystemErrorHandler)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java:84]: rename variable and field to fileSystemErrorHandler

File: presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java [hunk 15]
```diff
@@ -150,6 +155,7 @@ private void writePages(Iterator<Page> pageIterator)
             }
         }
         catch (UncheckedIOException | IOException e) {
+            this.filesystemErrorHandler.run();
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java:158]: - there is no point to use `this.`
- should we wrap this with try-catch so we don't swallow the current exception? If so please extract the method.

Comment (dfinninger) [reply presto-main/src/main/java/io/prestosql/spiller/FileSingleStreamSpiller.java:158]: Fixed usage of `this`, where applicable.

We could wrap the `fileSystemErrorHandler` in an exception, but the current impl is `LoadingCache::invalidateAll`, which doesn't throw an exception.

**Comment by kokosing** [comment]
Merged, thanks!
